Gun Accuracy Changes

comma opened the pull request:
First off, guns suck now.

Previously, when you missed, instead there was a 70% chance that you hit anyway, just different part. This effectively resulted in almost no misses no matter the accuracy.
Now that only happens for melee weapons (which it was added for in the first place IIRC).

'Aiming' changed a bit. At the moment if you stand still for few seconds, you get bonus to accuracy. This bonus is now capped by gun's accuracy.
Also 'aiming' time is not just moving now, every time you switch active hand away or tinker with your gun, that timer is reset (except for Master skill level).
Another important thing is that if you have no 'aiming' level (meaning you hadn't stood still for a second) you get a great penalty to accuracy depending on w_class of your gun.

Next, accuracy and miss chance calculations. The law has changed, now it's proportional to square of distance because linear wasn't giving me fast enough falloff.
Miss chance from distance is now decided by a bullet's variable, 'distance_falloff'. The higher it is, the faster miss chance grows with distance.

Guns also have new var 'accuracy_power' which decides how much accuracy bonuses affect the final miss chance.

Then I used new law and these coefs to muck up accuracy stuff for all guns. Now it works sorta like this.

Smol guns (low w_class, low two-handed penalty) are good at run&gun, if you're moving around a lot or switching hand away to do something, their accuracy won't suffer that much.
Bigger guns (rifles and such) are more accurate, but are rendered pretty useless without at least 1 second of aiming.
Within that, pistol rounds generally drop off accuracy fast with distance, meaning pistol-round guns will usually miss if fired further than 4 tiles.
Rifle rounds aren't as affected by distance, so they're good at firing from higher distances.
Lasers are rifle-tier, slightly worse for laser pistols just to maintain their close/far range roles. Think of it as shittier cooling or blurring or whatever realismo.

Another smol thing - moving now helps, moving around will give further miss chance (up to +10% miss chance), except when you move TOWARD the gun, then it's 0.25x of usual bonus. Standing still will instead give +5% chance to get hit.

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You'll find a README and example file in .\html\changelogs\ for further instructions.

You can also find a template for adding your changelog directly to the PR description here: https://github.com/Baystation12/Baystation12/wiki/Automatic-changelog-generation
-->

comma committed:
Gun Accuracy Changes

First off, guns suck now.

Previously, when you missed, instead there was a 70% chance that you hit anyway, just different part. This effectively resulted in almost no misses no matter the accuracy.
Now that only happens for melee weapons (which it was added for in the first place IIRC).

'Aiming' changed a bit. At the moment if you stand still for few seconds, you get bonus to accuracy. This bonus is now capped by gun's accuracy.
Also 'aiming' time is not just moving now, every time you switch active hand away or tinker with your gun, that timer is reset (except for Master skill level).
Another important thing is that if you have no 'aiming' level (meaning you hadn't stood still for a second) you get a great penalty to accuracy depending on w_class of your gun.

Next, accuracy and miss chance calculations. The law has changed, now it's proportional to square of distance because linear wasn't giving me fast enough falloff.
Miss chance from distance is now decided by a bullet's variable, 'distance_falloff'. The higher it is, the faster miss chance grows with distance.

Guns also have new var 'accuracy_power' which decides how much accuracy bonuses affect the final miss chance.

Then I used new law and these coefs to muck up accuracy stuff for all guns. Now it works sorta like this.

Smol guns (low w_class, low two-handed penalty) are good at run&gun, if you're moving around a lot or switching hand away to do something, their accuracy won't suffer that much.
Bigger guns (rifles and such) are more accurate, but are rendered pretty useless without at least 1 second of aiming.
Within that, pistol rounds generally drop off accuracy fast with distance, meaning pistol-round guns will usually miss if fired further than 4 tiles.
Rifle rounds aren't as affected by distance, so they're good at firing from higher distances.
Lasers are rifle-tier, slightly worse for laser pistols just to maintain their close/far range roles. Think of it as shittier cooling or blurring or whatever realismo.

Another smol thing - moving now helps, moving around will give further miss chance (up to +10% miss chance), except when you move TOWARD the gun, then it's 0.25x of usual bonus. Standing still will instead give +5% chance to get hit.

ghostsheet commented:
(I like this change? especially the movement part) but have you included rig suit mounted gun? is there an aiming time requirement for them?

The missing shot not having a chance to hit some other limbs kinda sucks, but I guess its fair? (I thought centre of mass was a thing though)

comma commented:
That aiming thing is for all guns, affected by their size. So it applies to mounted too, mounted eguns less affected, mountaed cannons more. Since they're not put in hands ever they're not affected by 'aim thrown off when you use hands to do something else' bit tho.

CakeQ commented on code/game/objects/items.dm:
```diff
@@ -785,6 +785,8 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 		. += "  <a href='?src=\ref[ID];look_at_id=1'>\[Look at ID\]</a>"
 
+/obj/item/proc/on_active_hand()
+
```
was this intentionally left blank?

comma replied:
yeah base item does nothing

CakeQ commented on code/modules/projectiles/gun.dm:
```diff
@@ -339,35 +337,36 @@
-			acc_mod -= 1
-		if(!held_twohanded)
-			acc_mod += -ceil(one_hand_penalty/2)
+		if(one_hand_penalty >= 4 && !held_twohanded)
```
can you not combine this with the if above?

CakeQ reviewed (commented):
Only other comment is that it might be a good idea later down the line to set up a projectiles define file with more consistent values (like distance_falloff_low, distance_falloff_high etc.)

afterthought2 commented on code/modules/projectiles/gun.dm:
```diff
@@ -339,35 +337,35 @@
-	var/stood_still = round((world.time - user.l_move_time)/15)
+	var/stood_still = started_aiming
+	//Not keeping gun active will throw off aim (for non-Masters)
+	if(user.l_move_time > started_aiming || user.skill_check(SKILL_WEAPONS, SKILL_PROF))
```
I think your skill check is inverted.

afterthought2 commented on code/modules/projectiles/gun.dm:
```diff
@@ -339,35 +337,35 @@
-			acc_mod += accuracy
+		acc_mod += min(max(2, accuracy), stood_still)
+	else if(w_class > ITEM_SIZE_NORMAL)
+		acc_mod -= 2*w_class - ITEM_SIZE_NORMAL
```
Is this really the intended order of operations, or did you intend `2 * (w_class - ITEM_SIZE_NORMAL)`?

comma replied:
Nope actually that.
I originally used two-handed penalty for run&gun penalties but it was getting out of hand how all the guns need it now, and how two-handed guns are bad at run&gun period (merc smg was affected by that)
So w_class was used instead, it punished big guns pretty harshly, and punishes pistols etc not that much.

afterthought2 replied:
OK, based on your explanation, do you mean it as is currently, or instead
```
else
	acc_mod -= w_class
	if(w_class > ITEM_SIZE_NORMAL)
		acc_mod -= w_class - ITEM_SIZE_NORMAL
```
Currently there is no penalty or bonus at ITEM_SIZE_NORMAL or below, but if you go over then you get a penalty of
```
ITEM_SIZE_NORMAL + 2 * (w_class - ITEM_SIZE_NORMAL)
```
On ITEM_SIZE_LARGE, which is 4, this adds a penalty of 5; on huge, it's 7. The sudden jump at NORMAL-> LARGE seemed like maybe not what you intended.

comma replied:
It's intended, but I'll be adjusting this with ANOTHER var soon, so might drop this thing.

afterthought2 replied:
OK, cool.

afterthought2 commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -144,15 +144,17 @@
-	return
+	var/obj/item/I = get_active_hand()
+	if(istype(I))
+		I.on_active_hand()
```
As far as I can tell, this proc will only be called when you swap hands, and in particular not when you remove the gun from the inventory/holster/turf and place it into your hand. Is this compatible with the rest of your changes, or do you need further handling?

comma replied:
Nah this proc is specifically for hand swapping, otherwise gun stuff uses attack hand.
I didn't consider holsters, should probably stick it in equipped instead.

afterthought2 commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -176,7 +176,7 @@ var/list/global/organ_rel_size = list(
 	miss_chance = max(miss_chance + miss_chance_mod, 0)
 	if(prob(miss_chance))
-		if(prob(70))
+		if(prob(70) || ranged_attack)
```
reverse the order

PurpleMartinJCK commented on code/modules/projectiles/gun.dm:
```diff
@@ -541,6 +538,10 @@
+	started_aiming = world.time
+
+/obj/item/weapon/gun/on_active_hand()
+	started_aiming = world.time
```
`started_aiming` is supposed to represent when it was last handled right? Makes me think that it counts the amount of "stood still" time that you mentioned in the main post. Perhaps name it something like `last_handled` or `became_active` or something.

comma committed:
Adds bulk var to guns, affecting rungun accuracy

w_class is decided for too many reasons unrealted to gun accuarcy so introduced another modifier to accuracy.

comma commented:
![](https://i.imgur.com/jy4Okwa.png)
the desired end result
